### PR TITLE
DAOS-6061 test: Fix random OSA online timeout.

### PR DESCRIPTION
**Pull request by rpadma2**
Summary :
     - threads array is initialized at the appropriate place.
     - Added the thread.join with timeout values.
     - Disable daos_racer by default (Seeing more rebuild failures with IOR/daos_racer in conjunction with OSA).
     - Move all the tests to run under medium clusters
     - Update YAML files to enable more debug messages (for rebuild).
     - Moved ior_thread common method to osa_utils.py
     - Check the rebuild status for failure condition.
     - Added job_manager_timeout to make sure IOR thread (hanging) is killed after 5 minutes
     - Skipped tests with correct ticket information.

Signed-off-by: Padmanabhan <ravindran.padmanabhan@intel.com>

### CODE REVIEW

**Review by shimizukko — APPROVED** [review]
LGTM

**Review by shimizukko — DISMISSED** [review]
LGTM

File: src/tests/ftest/util/osa_utils.py [hunk 1]
```diff
@@ -129,3 +147,52 @@ def verify_single_object(self):
...
+        Args:
+            pool (object): pool handle
+            oclass (str): IOR object class
+            api (str): IOR api
+            test (list): IOR test sequence
+            flags (str): IOR flags
+            results (queue): queue for returning thread results
+
+        """
+        processes = self.params.get("slots", "/run/ior/clientslots/*")
```
Comment (saurabhtandan) [inline src/tests/ftest/util/osa_utils.py:163]: minor: Can we have it consistent with IorTestBase class.
processes = self.params.get("np", '/run/ior/client_processes/*')

File: src/tests/ftest/util/osa_utils.py [hunk 2]
```diff
@@ -22,13 +22,17 @@
...
 from apricot import TestWithServers
 from command_utils import CommandFailure
+from ior_utils import IorCommand
+from job_manager_utils import Mpirun
+from mpio_utils import MpioUtils
 from pydaos.raw import (DaosContainer, IORequest,
                         DaosObj, DaosApiError)
 
-
 class OSAUtils(TestWithServers):
```
Comment (saurabhtandan) [inline src/tests/ftest/util/osa_utils.py:36]: How about Inheriting IorTestBase class here instead as that already inherits TestWithServers?

File: src/tests/ftest/util/osa_utils.py (hunk 1 above)
Comment (rpadma2) [reply src/tests/ftest/util/osa_utils.py:163]: Sure. I can do the changes.

File: src/tests/ftest/util/osa_utils.py (hunk 2 above)
Comment (rpadma2) [reply src/tests/ftest/util/osa_utils.py:36]: Sure.. Will give it a try...

File: src/tests/ftest/util/osa_utils.py (hunk 1 above)
Comment (rpadma2) [reply src/tests/ftest/util/osa_utils.py]: Done.

File: src/tests/ftest/util/osa_utils.py (hunk 2 above)
Comment (rpadma2) [reply src/tests/ftest/util/osa_utils.py]: Done.

**Comment by shimizukko** [comment]
Sorry. I'm on vacation, so not sure when I can do the review.

**Comment by rpadma2** [comment]
> Sorry. I'm on vacation, so not sure when I can do the review.

No problem Makito... We will take care of it here. Enjoy your vacation.

**Review by amjustin13 — APPROVED** [review]
Changes look good to me overall.

File: src/tests/ftest/osa/osa_offline_extend.py [hunk 3]
```diff
@@ -52,11 +52,9 @@ def run_offline_extend_test(self, num_pool, data=False):
         # Create a pool
         pool = {}
         pool_uuid = []
-        total_servers = len(self.hostlist_servers)
 
-        # Extend a rank (or server)
-        # rank index starts from zero
-        rank = total_servers
+        # Extend a ranks 4 and 5
```
Comment (saurabhtandan) [inline src/tests/ftest/osa/osa_offline_extend.py:56]: Minor: typo: Extend ranks 4 and 5

File: src/tests/ftest/osa/osa_online_drain.py [hunk 4]
```diff
@@ -189,43 +116,39 @@ def run_online_drain_test(self, num_pool):
                 for thrd in threads:
                     self.log.info("Thread : %s", thrd)
                     thrd.start()
-                    time.sleep(5)
+                    time.sleep(1)
```
Comment (saurabhtandan) [inline src/tests/ftest/osa/osa_online_drain.py:119]: hmm, hearing what Johann mentioned in one of out QA meetings, i think we should try to refrain from using sleep in our test code. Are you seeing any problem without sleep?

File: src/tests/ftest/osa/osa_online_extend.py [hunk 5]
```diff
@@ -150,16 +88,15 @@ def run_online_extend_test(self, num_pool):
         # Create a pool
         pool = {}
         pool_uuid = []
-        total_servers = len(self.hostlist_servers)
 
-        # Extend one of the ranks (or server)
-        # rank index starts from zero
-        rank = total_servers
+        # Extend one of the ranks 4 and 5
```
Comment (saurabhtandan) [inline src/tests/ftest/osa/osa_online_extend.py:92]: Minor: may be it should be
`Extend ranks 4 and 5`

File: src/tests/ftest/osa/osa_online_extend.py [hunk 6]
```diff
@@ -150,16 +88,15 @@ def run_online_extend_test(self, num_pool):
...
+        rank = [4, 5]
 
         # Start the daos_racer thread
-        daos_racer_thread = threading.Thread(target=self.daos_racer_thread)
-        daos_racer_thread.start()
-        time.sleep(30)
+        if racer is True:
+            daos_racer_thread = threading.Thread(target=self.daos_racer_thread)
+            daos_racer_thread.start()
+            time.sleep(30)
```
Comment (saurabhtandan) [inline src/tests/ftest/osa/osa_online_extend.py:99]: This is a big sleep. What is the intention of sleep here? Are we waiting for daos_racer to start or finish?
If waiting it to finish we can use daos_racer_thread.join()

File: src/tests/ftest/osa/osa_online_extend.py [hunk 7]
```diff
@@ -204,35 +141,31 @@ def run_online_extend_test(self, num_pool):
             self.log.info("Extra Servers = %s", self.extra_servers)
             self.start_additional_servers(self.extra_servers)
             # Give sometime for the additional server to come up.
-            time.sleep(5)
+            time.sleep(25)
```
Comment (saurabhtandan) [inline src/tests/ftest/osa/osa_online_extend.py:144]: 25 secs is again a bit long sleep.
Can we instead find the state of the newly added server if it has started or not? I see couple of methods in server_utils `detect_io_server_start()` and `get_single_system_state`. may be we can use one of those or if not that we can surely use dmg system query for the newly added rank and check it's state.

File: src/tests/ftest/osa/osa_online_reintegration.py [hunk 8]
```diff
@@ -181,9 +102,10 @@ def run_online_reintegration_test(self, num_pool):
         rank = random.randint(1, exclude_servers)
 
         # Start the daos_racer thread
-        daos_racer_thread = threading.Thread(target=self.daos_racer_thread)
-        daos_racer_thread.start()
-        time.sleep(30)
+        if racer is True:
+            daos_racer_thread = threading.Thread(target=self.daos_racer_thread)
+            daos_racer_thread.start()
+            time.sleep(30)
```
Comment (saurabhtandan) [inline src/tests/ftest/osa/osa_online_reintegration.py:108]: Same as above.

File: src/tests/ftest/util/osa_utils.py [hunk 9]
```diff
@@ -129,3 +180,51 @@ def verify_single_object(self):
...
+
+        """
+        container_info = {}
+        mpio_util = MpioUtils()
+        if mpio_util.mpich_installed(self.hostlist_clients) is False:
+            self.fail("Exiting Test : Mpich not installed on :"
+                      " {}".format(self.hostfile_clients[0]))
+        self.pool = pool
+        # Define the arguments for the ior_runner_thread method
+        ior_cmd = IorCommand()
```
Comment (saurabhtandan) [inline src/tests/ftest/util/osa_utils.py:203]: Can we use the self.ior_cmd param provided by IorTestBase?

**Review by saurabhtandan — CHANGES_REQUESTED** [review]

File: src/tests/ftest/osa/osa_offline_extend.py (hunk 3 above)
Comment (rpadma2) [reply src/tests/ftest/osa/osa_offline_extend.py:56]: Sure. I can fix it...

File: src/tests/ftest/osa/osa_online_drain.py (hunk 4 above)
Comment (rpadma2) [reply src/tests/ftest/osa/osa_online_drain.py:119]: Unfotunately, we need some timeouts here... I noticed ior jobs not started correctly... I am going to keep it right now.

File: src/tests/ftest/osa/osa_online_extend.py (hunk 5 above)
Comment (rpadma2) [reply src/tests/ftest/osa/osa_online_extend.py:92]: Sure will do.

File: src/tests/ftest/osa/osa_online_extend.py (hunk 7 above)
Comment (rpadma2) [reply src/tests/ftest/osa/osa_online_extend.py:144]: Good suggestion... I would like to create a new PR and will address this issue... With extend still blocked, let's wait for the actual dev changes before we can test this feature. I don't want to mess-up some thing with this PR... Already on my 24 CI run iteration... Really time consuming...

File: src/tests/ftest/util/osa_utils.py (hunk 9 above)
Comment (rpadma2) [reply src/tests/ftest/util/osa_utils.py:203]: No... I used this based on Samir's early suggestion that was not available via ior_cmd param. The skeleton of the code is same... Don't want to mess it now. Maybe later.

**Review by saurabhtandan — APPROVED** [review]
LGTM

File: src/tests/ftest/osa/osa_online_reintegration.py (hunk 8 above)
Comment (rpadma2) [reply src/tests/ftest/osa/osa_online_reintegration.py:108]: daos_racer and TestPool creation definitely needs some time difference... If not TestPool creation was failing... Di's doesn't want to use daos_racer tool... This was the reason, I have to stub it out...

File: src/tests/ftest/osa/osa_online_extend.py (hunk 6 above)
Comment (rpadma2) [reply src/tests/ftest/osa/osa_online_extend.py:99]: daos_racer runs in parallel during the whole OSA test execution... daos_racer creates pools and starts sending IO on a separate pool... While daos_racer initial setup is happening, creating a TestPool causes pool creation failures... That is the reason for the large timeout... Di is not fond of daos_racer tool running during OSA testing (due to bad IO data sent by the tool). To isolate the rebuild issues under OSA, we are temporarily stubbing out daos_racer tool.

**Review by amjustin13 — APPROVED** [review]
LGTM

**Comment by rpadma2** [comment]
daosgatekeeper, This PR is ready for merge.

**Pull request merged**